modules/encrypt/classes/Kohana/Encrypt/Engine/Sodium.php

```diff
@@ -0,0 +1,160 @@
+<?php
+
+/**
+ * The Encrypt Sodium engine provides two-way encryption of text and binary strings
+ * using the [Sodium](http://php.net/sodium) extension, which consists of two
+ * parts: the key and the cipher.
+ *
+ * The Key
+ * :  A secret passphrase that is used for encoding and decoding
+ *
+ * @package    Kohana/Encrypt
+ * @category   Security
+ * @author     Koseven Team
+ * @copyright  (c) Kohana Team
+ * @license    https://koseven.ga/LICENSE.md
+ */
+class Kohana_Encrypt_Engine_Sodium extends Kohana_Encrypt_Engine
+{
+	use Traits_Encrypt_Iv;
+
+	/**
+	 * @var string Engine type
+	 */
+	const TYPE = 'Sodium';
+
+	/**
+	 * @var int the size of the Initialization Vector (IV) in bytes
+	 */
+	protected $_iv_size;
+
+	/**
+	 * Sodium constructor.
+	 * @param array $config Array with configuration
+	 * @throws Kohana_Exception
+	 */
+	public function __construct($config)
+	{
+		parent::__construct($config);
+		if( ! extension_loaded('sodium') || ! sodium_crypto_aead_aes256gcm_is_available())
+		{
+			throw new Kohana_Exception('Sodium extension is not available');
+		}
+
+		$length = mb_strlen($this->_key, '8bit');
+
+		if ($length !== constant('SODIUM_CRYPTO_AEAD_AES256GCM_KEYBYTES'))
+		{
+			// No valid encryption key is provided!
+			throw new Kohana_Exception('No valid encryption key is defined in the encryption configuration: length should be :required_length for AEAD-AES-256-GCM, is: :current_length',[
+				':required_length' => constant('SODIUM_CRYPTO_AEAD_AES256GCM_KEYBYTES'),
+				':current_length' => $length
+			]);
+		}
+
+		$this->_iv_size = constant('SODIUM_CRYPTO_AEAD_AES256GCM_NPUBBYTES');
+	}
+
+	/**
+	 * Encrypts
+	 * @param string $message Your message to be encrypted
+	 * @param string $iv
+	 * @return null|string
+	 */
+	public function encrypt(string $message, string $iv)
+	{
+		$value = sodium_crypto_aead_aes256gcm_encrypt($message, '', $iv, $this->_key);
+
+		if ($value === FALSE)
+		{
+			// Encryption failed
+			return NULL;
+		}
+
+		//Base64 encode binary data, otherwise they cannot be
+		//transformed into JSON.
+		$value = base64_encode($value);
```

Comment:
This is unnecessary according to documentation https://github.com/paragonie/pecl-libsodium-doc/blob/master/chapters/08-advanced.md

just base64_encode($value) is more than enough.